### PR TITLE
Batch API Commands

### PR DESCRIPTION
**Pull request by blinktag**
This PR adds the following commands:

| Command     | Description |
| ----------- | ----------- |
| `plumber batch login` | Login and save credentials to ~/.batchsh/config.json |
| `plumber batch logout` | Logout and clear saved credentials |
| `plumber batch list collection` | List collections |
| `plumber batch create collection --name "My events" --schema-id "0e31f330-aca0-468c-a186-a26e889a19e9"` | Create collection |
| `plumber batch search --collection-id="..." --query="*"` | Search a collection |
| `plumber batch list schema` | List schemas |
| `plumber batch list replay` | List replays |
| `plumber batch list destination` | List destinations | 
| `plumber batch create destination kafka` | Create kafka destination |
| `plumber batch create destination aws-sqs` | Create SQS destination |
| `plumber batch create destination rabbit` | Create RabbitMQ destination |
| `plumber batch create destination http` | Create HTTP destination |
| `plumber batch create replay` | Create and start a new replay |

### CODE REVIEW

**Comment by dselans** [comment]
The current command format is `plumber ACTION something` - should we do the same for the batch commands? Ie. `plumber batch create collection ...`, `batch list collection ...`, `batch create replay ...` , `batch list replay ...`, etc.

**Comment by dselans** [comment]
Also, we should have the ability to create replays and destinations via cli.

File: backends/batch/replays.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+func (b *Batch) pauseReplay() error {
+	// TODO
+
+	return nil
+}
+
+func (b *Batch) resumeReplay() error {
+	// TODO
```
Comment (dselans) [inline backends/batch/replays.go:100]: If these get called somewhere - let's add a `logrus.Warning("resume not implemented")`.

File: backends/batch/replays.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+	return nil
+}
+
+func (b *Batch) generateReplayQuery() (string, error) {
+	from, err := time.Parse("2006-01-02T15:04:05Z", b.Opts.Batch.ReplayFrom)
+	if err != nil {
+		return "", fmt.Errorf("--from-timestamp '%s' is not a valid RFC3339 date", b.Opts.Batch.ReplayFrom)
+	}
+
+	to, err := time.Parse("2006-01-02T15:04:05Z", b.Opts.Batch.ReplayFrom)
```
Comment (dselans) [inline backends/batch/replays.go:163]: `b.Opts.Batch.ReplayTo`

File: cli/batch.go [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+	cmd.Flag("query", "Search query").
+		Required().
+		StringVar(&opts.Batch.Query)
+
+	cmd.Flag("from-timestamp", "RFC3339 timestamp for where to begin event search query").
+		Required().
+		StringVar(&opts.Batch.ReplayFrom)
+
+	cmd.Flag("to-timestamp", "RFC3339 timestamp for where to send event search query").
```
Comment (dselans) [inline cli/batch.go:120]:
```suggestion
	cmd.Flag("to-timestamp", "RFC3339 timestamp for where to end event search query").
```

**Review by dselans — APPROVED** [review]
That's a lot of code. All minor comments for small fixes. Let me run this locally before you merge.

File: backends/batch/replays.go (hunk 1 above)
Comment (blinktag) [reply backends/batch/replays.go:100]: Doesn't get called anywhere yet

File: backends/batch/config.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	_, err = f.Write(data)
+	return err
+}
+
+// getConfigDir returns a directory where the batch configuration will be stored
+func getConfigDir() (string, error) {
+	// Get user's home directory
+	usr, err := user.Current()
```
[the remark this answers is not in the document]
Comment (blinktag) [reply backends/batch/config.go:109]: Changed to https://golang.org/pkg/os/#UserHomeDir which supports Windows without issue

**Pull request merged**